File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    var answer = 0;
    
    const graph = Array.from({length: n+1}, ()=> []);
    
    for(const [a,b] of edge){
        graph[a].push(b);
        graph[b].push(a);
    }
    
    const dist = Array.from({length: n+1}, ()=> -1);
    const q = []
    
    dist[1] = 0;
    q.push(1);
    
    while(q.length){
        const cur = q.shift();
        
        for(const next of graph[cur]){
            if(dist[next] !== -1)   continue;
            dist[next] = dist[cur] + 1;
            q.push(next);
        }
    }
    
    const max = Math.max(...dist);
    
    for(let i=0; i<dist.length;i++){
        if(dist[i]===max)   answer++;
    }
    
    
    return answer;
}

// 최단 거리? -> BFS
// DFS는 길어도 계속 가서 최단거리 보장 X

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque
from collections import defaultdict

def solution(n, edge):
    answer = 0
    
    nodes = defaultdict(int)
    graph = defaultdict(list)    

    visited = [False] * (n+1)
    queue = deque()
    
    nodes[1] = 0
    visited[1] = True
    queue.append((1, 0))
    
    for e in edge:
        start, end = e
        graph[start].append(end)
        graph[end].append(start)
    
    while queue:
        cur, cnt = queue.popleft()
        
        for next in graph[cur]:
            if not visited[next]:
                visited[next] = True
                nodes[next] = cnt + 1
                queue.append((next, cnt+1))
    
    max_node = max(nodes.values())
    
    for node in nodes.values():
        if node == max_node:
            answer += 1
            
    return answer